Client/TabManagement/TabManagerImplementation.swift

```diff
@@ -467,3 +475,75 @@ class TabManagerImplementation: LegacyTabManager, Notifiable {
         }
     }
 }
+
+extension TabManagerImplementation {
+    // Ecosia: Get Sites from places
+    private func fetchDBSites() async -> [Site] {
+        do {
+            let sites = try await withCheckedThrowingContinuation { continuation in
+                profile.places.getSitesWithBound(
+                    limit: 100,
+                    offset: 0,
+                    excludedTypes: VisitTransitionSet(0)
+                ).upon { result in
+                    if let successValue = result.successValue {
+                        continuation.resume(returning: successValue.asArray())
+                    } else {
+                        continuation.resume(returning: [])
+                    }
+                }
+            }
+            return sites
+        } catch {
+            return []
+        }
+    }
+    
+    // Ecosia: Fix tabs that where migrated from v9.x to v10.0.0 by fetching from places DB instead.
+    // This can be removed after there are no significant number of users on v10.0.0.
+    private static let restoreMigratedv10TabsMissingUrlKey = "restoreMigratedv10TabsMissingUrl"
+    private var shouldRestoreMigratedv10TabsMissingUrl: Bool {
+        !UserDefaults.standard.bool(forKey: Self.restoreMigratedv10TabsMissingUrlKey)
+    }
+    private func restoreMigratedv10TabsMissingUrlIfNeeded(window: WindowData?) async -> WindowData? {
+        defer {
+            UserDefaults.standard.setValue(true, forKey: Self.restoreMigratedv10TabsMissingUrlKey)
+        }
+        // We only want to run it if the user just upgraded to the version containing this restoration code.
+        // Otherwise it means the user is a fresh install and we just mark it as restoration done.
+        guard EcosiaInstallType.get() == .upgrade else {
+            return window
+        }
+        // We only restore if there are actually any tabs with empty url.
+        guard let window = window,
+              window.tabData.filter({ $0.siteUrl.isEmpty }).count > 0 else {
+            return window
+        }
+        let sites = await fetchDBSites()
+        var restoredTabs = [TabData]()
+        window.tabData.forEach { tab in
+            var restoredUrl = tab.siteUrl
+            if restoredUrl.isEmpty {
+                restoredUrl = sites.first(where: { $0.title == tab.title })?.url ?? ""
+                // If we don't have a URL and the tab has no title, it means it should be the homepage (or at least that's better than blank)
+                if restoredUrl.isEmpty && tab.title?.isEmpty == true {
+                    restoredUrl = "internal://local/about/home"
```

Comment:
is there a constant we could use instead of the plain string?

Reply:
I wish there was!
I searched thru the codebase and couldn't find anything.
We can make one ourselves perhaps 🤷 and use it for our own things without changing Firefox's ones

Reply:
Added a constant and used on all places needed, turns out it wasn't that much.

Reply:
@lucaschifino I wonder whether it's safe to change that for Firefox's too... of course we can see that in conflicts when upgrading but wondering whether that page would change in its URL structure from the Firefox's side.
I'm happy with having it for the time being...just raising a 💭 out loud

Reply:
If it changes on Firefox side then we need to change it in ours too, so I think this is a good approach explicitly appearing on future conflicts. Even so, it might be we'll want to revert this PR in the future after restoring is no longer valid and in that case we can revert this with it (since we'll no longer be using the homepage url and don't care if it changes).